modules/shared/src/main/scala/org/tessellation/currency/dataApplication/storage/CalculatedStateLocalFileSystemStorage.scala

```diff
@@ -0,0 +1,57 @@
+package org.tessellation.currency.dataApplication.storage
+
+import cats.Applicative
+import cats.effect.kernel.Async
+import cats.syntax.all._
+
+import org.tessellation.schema.SnapshotOrdinal
+import org.tessellation.storage.LocalFileSystemStorage
+
+import eu.timepit.refined.types.numeric.NonNegLong
+import fs2.Stream
+import fs2.io.file.Path
+
+final class CalculatedStateLocalFileSystemStorage[F[_]: Async] private (
+  path: Path
+) extends LocalFileSystemStorage[F, Array[Byte]](path) {
+  def read[A](ordinal: SnapshotOrdinal)(implicit decoder: Array[Byte] => F[A]): F[Option[A]] =
+    readBytes(toOrdinalName(ordinal)).flatMap {
+      _.traverse(decoder)
+    }
+
+  def exists(ordinal: SnapshotOrdinal): F[Boolean] =
+    exists(toOrdinalName(ordinal))
+
+  def write[A](ordinal: SnapshotOrdinal, state: A)(implicit encoder: A => F[Array[Byte]]): F[Unit] =
+    encoder(state).flatMap(write(toOrdinalName(ordinal), _))
+
+  def delete[A](ordinal: SnapshotOrdinal): F[Unit] =
+    delete(toOrdinalName(ordinal))
+
+  def listStoredOrdinals: F[Stream[F, SnapshotOrdinal]] =
+    listFiles.map {
+      _.map(_.name)
+        .map(_.toLongOption)
+        .map(_.flatMap(NonNegLong.from(_).toOption).map(SnapshotOrdinal(_)))
+        .collect {
+          case Some(a) => a
+        }
+    }
+
+  def deleteAbove(ordinal: SnapshotOrdinal): F[Unit] =
+    listStoredOrdinals.flatMap {
+      _.filter(_ > ordinal)
+        .evalMap(delete(_))
+        .compile
+        .drain
+    }
+
```

Comment:
Most of these methods are exactly the same as the ones in `SnapshotInfoLocalFileSystemStorage`. Maybe it is worth to move common methods to base class or create a mixin like `OrdinalLocalFileSystemStorage`

Reply:
Also, this will be the third instance of converting a long to `SnapshotOrdinal`. Our code will be a bit cleaner if we implement this in the companion object of `SnapshotOrdinal`:
```
def from(value: Long): Option[SnapshotOrdinal] =
  NonNegLong.from(value).toOption.map(SnapshotOrdinal(_))
```

Reply:
:+1: